crates/curp/src/server/storage/wal/segment.rs

```diff
@@ -0,0 +1,545 @@
+use std::{io, iter, pin::Pin, sync::Arc, task::Poll};
+
+use clippy_utilities::{NumericCast, OverflowArithmetic};
+use curp_external_api::LogIndex;
+use futures::{ready, FutureExt, SinkExt};
+use serde::{de::DeserializeOwned, Serialize};
+use tokio::{
+    fs::File as TokioFile,
+    io::{AsyncRead, AsyncReadExt, AsyncSeekExt, AsyncWrite, AsyncWriteExt},
+    sync::Mutex,
+};
+use tokio_stream::StreamExt;
+use tokio_util::codec::Framed;
+
+use crate::log_entry::LogEntry;
+
+use super::{
+    codec::{DataFrame, WAL},
+    error::{CorruptType, WALError},
+    util::{get_checksum, parse_u64, validate_data, LockedFile},
+    WAL_FILE_EXT, WAL_MAGIC, WAL_VERSION,
+};
+
+/// The size of wal file header in bytes
+const WAL_HEADER_SIZE: usize = 56;
+
+/// A segment of WAL
+#[derive(Debug)]
+pub(super) struct WALSegment {
+    /// The base index of this segment
+    base_index: LogIndex,
+    /// The id of this segment
+    segment_id: u64,
+    /// The soft size limit of this segment
+    size_limit: u64,
+    /// The opened file of this segment
+    file: TokioFile,
+    /// The file size of the segment
+    size: u64,
+    /// The highest index of the segment
+    seal_index: LogIndex,
+    /// The IO state of the file
+    io_state: IOState,
+}
+
+/// The IO state of the file
+#[derive(Clone, Copy, Debug, Default)]
+pub(super) enum IOState {
+    /// The initial state that haven't written any data or fsynced
```

Comment:
It looks weird. Does a state that haven't written any data is identical to a fsynced state? I think they're different